src/proxy/resolve.rs

```diff
@@ -160,8 +190,74 @@ where
     }
 }
 
+// === impl EndpointStatus ===
+
 impl EndpointStatus {
     pub fn is_empty(&self) -> bool {
         self.0.load(Ordering::Acquire)
     }
 }
+
+// === impl MakeStream ===
+
+impl<F> MakeStream<F> {
+    fn push(&mut self, addr: SocketAddr, inner: F) {
+        let (cancel, canceled) = oneshot::channel();
+        self.cancellations.insert(addr, cancel);
+        self.futures.push(MakeFuture {
+            addr,
+            inner,
+            canceled,
+        });
+    }
+
+    fn remove(&mut self, addr: &SocketAddr) {
+        if let Some(cancel) = self.cancellations.remove(addr) {
+            let _ = cancel.send(());
+        }
+    }
+}
+
+impl<F: Future> Stream for MakeStream<F> {
+    type Item = (SocketAddr, F::Item);
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            return match self.futures.poll() {
+                Err(MakeError::Canceled) => continue,
+                Err(MakeError::Inner(err)) => Err(err),
+                Ok(Async::Ready(Some((addr, svc)))) => {
+                    let _rm = self.cancellations.remove(&addr);
+                    debug_assert!(_rm.is_some(), "cancellation missing");
```

Comment:
TIOLI: since we have the address anyway, maybe
```suggestion
                    debug_assert!(_rm.is_some(), "cancellation missing for {:?}", addr);
```
would be worth including? could make debugging easier if we ever actually hit this assertion...